columnflow/production/cms/btag.py

```diff
@@ -6,18 +6,60 @@
 
 from __future__ import annotations
 
+from dataclasses import dataclass, field
+
 import law
 
 from columnflow.production import Producer, producer
 from columnflow.util import maybe_import, InsertableDict
 from columnflow.columnar_util import set_ak_column, flat_np_view, layout_ak_array
+from columnflow.types import Any
 
 np = maybe_import("numpy")
 ak = maybe_import("awkward")
 
 logger = law.logger.get_logger(__name__)
 
 
+@dataclass
+class BTagSFConfig:
+    correction_set: str
+    jec_sources: list[str]
+    discriminator: str = ""  # when empty, set in post init based on correction set
+    corrector_kwargs: dict[str, Any] = field(default_factory=dict)
+
+    def __post_init__(self):
+        cs = self.correction_set.lower()
+        if not self.discriminator:
+            if "deepjet" in cs:
+                self.discriminator = "btagDeepFlavB"
+            elif "particlenet" in cs:
+                self.discriminator = "btagPNetB"
+            else:
+                raise NotImplementedError(
+                    "cannot identify btag discriminator for correction set "
+                    f"'{self.correction_set}', please set it manually",
+                )
+
+        # warn about potentially wrong column usage
+        if (
+            ("deepjet" in cs and "pnet" in self.discriminator) or
+            ("particlenet" in cs and "deepflav" in self.discriminator)
+        ):
+            logger.warning(
+                f"using btag column '{self.discriminator}' for btag sf corrector "
+                f"'{self.correction_set}' is highly discouraged",
+            )
+
+    @classmethod
+    def new(
+        cls,
+        obj: BTagSFConfig | tuple[str, list[str]] | tuple[str, list[str], str],
+    ) -> BTagSFConfig:
+        # purely for backwards compatibility with the old tuple format
+        return obj if isinstance(obj, cls) else cls(*obj)
```

Comment:
Could we also allow passing a simple dictionary?

Reply:
In principle yes, but thinking about it, I would prefer these dataclass-style container objects since they allow for some build-in parsing and validation. In the long-term I would even remove the backwards compatibility and no longer accept tuples.

Did you have a special use case in mind that would require dicts?

Reply:
Not really, just laziness such that a user does not need to import the `BTagSFConfig` for their own analysis as the checks from the dataclass object runs anyways when the `get_btag_config` function is called in the init.